classes/class3/catscatscats/routes/mg_funcs.js

```diff
@@ -0,0 +1,121 @@
+var mongoose = require('mongoose');
+
+console.log(process.env.MONGORURI);
+var mongoURI = process.env.MONGOURI || "mongodb://catsuser:jscats@ds039411.mongolab.com:39411/olinjs";
+mongoose.connect(mongoURI);
+
+var db = mongoose.connection;
+db.on('error', console.error.bind(console, 'connection error:'));
+db.once('open', function (callback) {
+  console.log('Successfully opened mongo database: ', mongoURI);
+});
+
+var catSchema = mongoose.Schema({
+  name: String,
+  age: Number,
+  colors: [String]
+});
+var Cat = mongoose.model('Cat', catSchema);
+
+var firstNames = ['Hagar', 'Catface', 'Hardcastle',
+                  'Harmony', 'Ronnie', 'Harry',
+                  'Professor', 'Upfish', 'Valmart',
+                  'Dazzler', 'Roast-Beefy', 'Joey'];
+var lastNames  = ['the Horrible', 'Meowmers', 'McCormick',
+                  'the Wretch', 'the Bear', 'Freaking Potter',
+                  'Snake', 'Hufferpuffer', 'Viacom',
+                  'the Half-Dead', 'O\'Weefy', 'Lumbermouth'];
+var colors = ['red', 'yellow', 'blue', 'invisible', 'tangerine',
+              'poopy brown', 'lime', 'chartreuse', 'olive',
+              'seafoam', 'aquamarine', 'azure', 'puce', 'puke'];
+
+function choose (choices) {
+  var index = Math.floor(Math.random() * choices.length);
+  return choices[index];
+}
+
+function pickColors() {
+  var res = [];
+  colors.forEach(function(col) {
+    if (Math.random() > 0.8) {
+      res.push(col);
+    }
+  });
+  if (res.length == 0) {
+    res.push(choose(colors));
+  }
+  return res;
+}
+
+var addCat = function() {
+  var catName = choose(firstNames) + ' ' + choose(lastNames);
+  var catColors = pickColors();
+  var catAge = Math.floor(Math.random() * 18) + 1;
+
+  var cat = new Cat({name: catName, age: catAge, colors: catColors});
+
+  cat.save(function (err) {
+    if (err) {
+      console.log("Problem saving cat", err);
+    }
+  });
+
+  return catToString(cat);
+}
+
+function catToString(cat) {
+  return cat.name + 
+         ' the ' +
+         cat.age.toString() +
+         '-year-old, '
+         + cat.colors.join(' and ') +
+         ' colored cat';
+}
+
+function catsByColor(res, color) {
+  console.log("HELLO" + color);
+  Cat.find( { colors: color } )
+    .exec(function (err, cats) {
+      // Once the cats have been found,
+      // render them.
+      cats = cats.map(catToString);
```

Comment:
Nice use of map, however you should be doing this formatting in a template, not as a toString method.